File: src/01_data_structures/05_hash_table/implementations/ts/hashTableLinkedList.ts
import { LinkedList } from "../../../02_linked_list/implementations/ts/linked_list";

/*
  LINKED LIST FOR HASH TABLES
  In hash tables we will store items in our linked lists as key, value tuples.
  To find an item by the key we need a method that receives a matcher function, and will return the item that matches it.
  Thus, we extend the Linked list to add this behavior.
*/
interface IHashTableLinkedList<T> extends LinkedList<T> {
  getItemByCallback(matcher: (arg: T) => boolean): T | null;
}

export class HashTableLinkedList<T>
  extends LinkedList<T>
  implements IHashTableLinkedList<T>
{
  getItemByCallback(matcher: (arg: T) => boolean): T | null {
    let current = this.head;

    while (current !== null) {
      // If found
      if (matcher(current.value)) {
        return current.value;
      }
      current = current.next;
    }

    // Not found
    return null;
  }
}

File: src/01_data_structures/05_hash_table/implementations/go/hash_table_linked_list.go
package hash_table

import linkedList "data-structures-algorithms-problems/src/01_data_structures/02_linked_list/implementations/go"

type HashTableLinkedListItem[T comparable] struct {
	Key   string
	Value T
}

type IHashTableLinkedList[T comparable] interface {
	Prepend(hashTableLinkedListItem HashTableLinkedListItem[T]) IHashTableLinkedList[T]
	Append(hashTableLinkedListItem HashTableLinkedListItem[T]) IHashTableLinkedList[T]
	Remove(matcher func(HashTableLinkedListItem[T]) bool) IHashTableLinkedList[T]
	Find(matcher func(HashTableLinkedListItem[T]) bool) bool
	GetItemIf(matcher func(HashTableLinkedListItem[T]) bool) (T, bool)
}

type HashTableLinkedList[T comparable] struct {
	LinkedList *linkedList.LinkedList[HashTableLinkedListItem[T]]
}

func (this *HashTableLinkedList[T]) Prepend(hashTableLinkedListItem HashTableLinkedListItem[T]) IHashTableLinkedList[T] {
	this.LinkedList.Prepend(hashTableLinkedListItem)

	return this
}

func (this *HashTableLinkedList[T]) Append(hashTableLinkedListItem HashTableLinkedListItem[T]) IHashTableLinkedList[T] {
	this.LinkedList.Append(hashTableLinkedListItem)

	return this
}

func (this *HashTableLinkedList[T]) Remove(matcher func(HashTableLinkedListItem[T]) bool) IHashTableLinkedList[T] {
	this.LinkedList.Remove(matcher)

	return this
}

func (this *HashTableLinkedList[T]) Find(matcher func(HashTableLinkedListItem[T]) bool) bool {
	return this.LinkedList.Find(matcher)
}

func (this *HashTableLinkedList[T]) GetItemIf(matcher func(HashTableLinkedListItem[T]) bool) (T, bool) {
	var zero T

	current := this.LinkedList.Head

	for current != nil {
		if matcher(current.Value) {
			return current.Value.Value, true
		}

		current = current.Next
	}

	return zero, false
}
